(Preceding snippet PR) Improved explanation of Wait/Ticker/Timeout/Threads/EventQueue (#1320)
ARMmbed/Handbook: docs/api/platform/flow_control.md

# Application flow control

We can use Blinky to explore flow control and task management in Arm Mbed OS applications. We'll look at automating actions with delays first, then move on to handling user actions.

## Automating actions with delays

If we want to automatically blink an LED, we have four main techniques:

1. [Wait API](#wait-api)
1. [Ticker and Timeout](#ticker-and-timeout)
1. [Thread](#thread)
1. [EventQueue](#eventqueue)

The techniques cater to different requirements for precision, efficiency and context.

### Wait API

The Wait API creates delays by actively blocking the execution for a period of time. It uses the busy wait strategy - internally running a loop until the end of the period.

> Here is an example that uses `wait_us()` of the API to wait a number of microseconds:

Reviewer: ```suggestion
Here is an example that uses the API's `wait_us()` function to wait a number of microseconds:
```

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait/blob/v6.0/main.cpp)

> The processor is active for the entire duration, so the Wait API is suitable when you need short delays (nanoseconds to a few microseconds) without triggering [sleep modes](../apis/power-management-sleep.html) which have some wake-up latencies. It is also safe to use in interrupt contexts. However, we do not recommend busy waiting for longer delays, during which the system may not be able to switch to other tasks or save power as it cannot sleep.

Reviewer: ```suggestion
The processor is active for the entire duration, so the Wait API is suitable when you need short delays (nanoseconds to a few microseconds) without triggering [sleep modes](../apis/power-management-sleep.html), which have some wake-up latencies. It is also safe to use in interrupt contexts. However, we do not recommend busy waiting for longer delays: the system may not be able to switch to other tasks or save power by sleeping.
```

Reviewer: I like that!

Here is an example that uses a ticker object:

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker/blob/v6.0/main.cpp)

<span class="warnings"> **Warning:** A ticker/timeout's handlers are executed in interrupt contexts and thus, like any interrupt handlers, should return quickly and not use `printf` or APIs that are not intended for interrupts.</span>

If you don't need the precision of a high-frequency Ticker or Timeout, we recommend that you use [LowPowerTicker](../apis/lowpowerticker.html) or [Low Power Timeout](../apis/lowpowertimeout.html) instead. The low power classes inform the operating system you want to allow deep sleep mode on your system. Note that entering deep sleep also depends on the specific environment and characteristics of your system, not just your API selection. For more information about sleep and deep sleep, please refer to our [documentation about power management](../apis/power-management-sleep.html) and our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s).

### Thread

> If your application is running in RTOS mode then [Threads](../apis/thread.html) are another efficient way to blink an LED. During the waiting period, it is possible to take advantage of Mbed OS optimizations to automatically conserve power and deal with other tasks. This makes Threads the most efficient way to run tasks in Mbed OS.

Reviewer: So this is my fault - in my previous edit, I created the impression that you have to *manually* use the Mbed OS optimizations. I'll change it.

Reviewer: ```suggestion
If your application is running in RTOS mode then [Threads](../apis/thread.html) are another efficient way to blink an LED. During the waiting period, Mbed OS automatically conserves power and deals with other tasks. This makes Threads the most efficient way to run tasks in Mbed OS.
```

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread/blob/v6.0/main.cpp)

### EventQueue

The [EventQueue](../apis/eventqueue.html) class uses the `call_every()` function to schedule repeated actions:

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue/main.cpp)

> For one-off delays, use `call_in()`. Just as with Ticker and Timeout, if no threads are running during a wait, the system enters [sleep or deep sleep mode](../apis/power-management-sleep.html). A major advantage of EventQueue over Ticker and Timeout is that the handler is called in the same context as the EventQueue is dispatched (thread, in the case of RTOS), so interrupt-related restrictions (such as no `printf` oand no `Mutex`) do not apply.

Reviewer: ```suggestion
For one-off delays, use `call_in()`. Just as with Ticker and Timeout, if no threads are running during a wait, the system enters [sleep or deep sleep mode](../apis/power-management-sleep.html). A major advantage of EventQueue over Ticker and Timeout is that the handler is called in the same context as the EventQueue is dispatched (thread, in the case of RTOS), so interrupt-related restrictions (such as no `printf` and no `Mutex`) do not apply.
```

We can use Blinky to explore flow control and task management in Arm Mbed OS applications. We'll look at automated actions first, then move on to handling user actions.

> ## Flow control for delayed actions

Reviewer: This changed from "automated" to "delayed" but the other one is still "manual". So now the contrast isn't as clear to me. 

Also, line 3 uses "automated" and "user", which doesn't match either header.

Author: Good question. I changed "automated" because it sounded too vague/wide as the section is just above delays.

Personally I think "Automating actions with delays" and "Handling user actions" are clearer titles - I'll also update the intro in line 3 for consistency.

If we want to automatically blink an LED, we have four main techniques:

1. [Busy wait](#busy-wait)
1. [Ticker, Timeout](#ticker-timeout)
1. [Thread](#thread)
1. [EventQueue](#eventqueue)

> All above techniques accomplish delays but cater to different requirements for precision, efficiency and context.

Reviewer: ```suggestion
The techniques cater to different requirements for precision, efficiency and context.
```

> <span class="tips">**Tip:** You may want to read the [power optimization](../apis/platform-concepts.html) tutorial to understand how to achieve power savings. </span>

Reviewer: ```suggestion
<span class="tips">**Tip:** You may want to read the [power optimization](../apis/platform-concepts.html) tutorial to learn how to save power. </span>
```

Reviewer: Is this related specifically to automated/delayed actions, or should it be before this section (as a generic "flow control is related to power optimization and you should read both")?

Author: I think "Further readings" would be a better place. It's not a strict prerequisite, and this tutorial discusses a bit around power optimization along the way. Putting the link at the end minimises distraction in my opinion.

### Busy wait

> Busy wait is a method that blocks the processor for a period of time. The processor runs at full power for the duration of the wait:

Reviewer: When will I want to use this?
Is it running at full power because it's working on something else? Or just because busy wait doesn't let it power down?

Author: Good point, now I'm discussing with the team how to better explain this.

Author: I've just updated the explanation

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait/blob/v6.0/main.cpp)

> Notice `wait_us()` - it is part of the [Wait API](../mbed-os-api-doxy/group__platform__wait__api.html) to busy wait a given number of microseconds.

Reviewer: ```suggestion
Notice `wait_us()` - it is part of the [Wait API](../mbed-os-api-doxy/group__platform__wait__api.html), and sets the busy wait time as number of microseconds.
```

> The Wait API is an ISR-safe way to create short delays of nanoseconds to a few microseconds. However, we do not recommend busy waiting for longer delays.

Reviewer: What's ISR?

Reviewer: Again, because I'm not sure what I'm trying to achieve with these delays, I don't know why I'll want to use busy wait.

Reviewer: This is talking about the Wait API and busy wait as if they're one and the same. Is this true?

Author: Will improve this

> The following techniques are better suited for milliseconds to seconds delays which apply to our example's use case.

Reviewer: ```suggestion
For longer delays - millisecond to seconds - use one of the other techniques. 
```

> ### Ticker, Timeout

Reviewer: ```suggestion
### Ticker and Timeout
```

Tickers and Timeouts are non-blocking, interrupt-based ways of creating a time interval - your code continues to execute or sleep when there is nothing to do. The difference is that Tickers are recurring whereas Timeouts are one-off.

Here is an example that uses a ticker object:

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker/blob/v6.0/main.cpp)

<span class="warnings"> **Warning:** A ticker/timeout's handlers are executed in ISR context and thus, like any ISR handlers, should return quickly and not use `printf` or APIs that are not intended for ISRs.</span>

If you don't need the precision of a high-frequency ticker or timeout, we recommend that you use LowPowerTicker or LowPowerTimeout instead. These allow the system to be put in deep sleep mode.

> - _High-resolution microsecond ticker/timeout_ ([Ticker](ticker.html), [Timeout](timeout.html))

Reviewer: These two line should go at the end of this section, as "further reading".

Author: I converted them to inline links

> - _High-resolution microsecond ticker/timeout_ ([Ticker](ticker.html), [Timeout](timeout.html))

Reviewer: ```suggestion
- High-resolution microsecond [Ticker](../apis/ticker.html) and [timeout](../apis/timeout.html). 
```

> - _Low Power ticker/timeout_ ([LowPowerTicker](lowpowerticker.html), [LowPowerTimeout](lowpowertimeout.html))

Reviewer: ```suggestion
- [Low Power Ticker](../apis/lowpowerticker.html) and [Low Power Timeout](../apis/lowpowertimeout.html).
```

> Usage of the low power classes will inform the operating system of your desire to allow Deep Sleep mode on your system, although actually entering deep sleep depends on the specific environment and characteristics of your system. For more information about Sleep and Deep Sleep, please refer to our [documentation page](power-management-sleep.html) as well as watch our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s) where the concepts of Sleep and Deep Sleep are described in-depth.

Reviewer: ```suggestion
Use the the low power classes to inform the operating system you want to allow deep sleep mode on your system. Note that entering deep sleep depends on the specific environment and characteristics of your system, not just your API selection. For more information about sleep and deep sleep, please refer to our [documentation about power management](../apis/power-management-sleep.html) and our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s).
```

### Thread

> If your application is running in RTOS mode then [Thread](../apis/thread.html)s are another efficient way to blink an LED. During the waiting period, it is possible to take advantage of Mbed OS optimizations to automatically conserve power and deal with other tasks.

Reviewer: ```suggestion
If your application is running in RTOS mode then [Threads](../apis/thread.html) are another efficient way to blink an LED. During the waiting period, it is possible to take advantage of Mbed OS optimizations to automatically conserve power and deal with other tasks. This makes Threads the most efficient way to run tasks in Mbed OS.
```

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread/blob/v6.0/main.cpp)

> Threads are the most efficient ways to run tasks in Mbed OS since the operating system internally contains optimizations for efficient scheduling of tasks and maximizing the system's sleep time.

Reviewer: ```suggestion
```

Reviewer: We already mentioned this in the previous paragraph. Would be good to combine those, so I'm deleting this and putting the info there.

### EventQueue

> The [EventQueue](../apis/eventqueue.html) class has a simple-to-use `call_every()` function to schedule repeated actions:

Reviewer: ```suggestion
The [EventQueue](../apis/eventqueue.html) class uses the `call_every()` function to schedule repeated actions:
```

Reviewer: We don't tell users something is simple to use. That's for them to decide.

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue/main.cpp)

> <span class="tips">**Tip:** For one-off delays, use `call_in()`.

Reviewer: ```suggestion
For one-off delays, use `call_in()`.
```

Reviewer: Need to remove the line breaks between these three paragraphs - they're a single explanation and should be a single paragraph.

> Just as with Ticker/Timeout, if no threads are running during a wait, the system enters sleep mode.</span>

Reviewer: ```suggestion
Just as with Ticker and Timeout, if no threads are running during a wait, the system enters sleep mode.
```

Reviewer: Sleep or deep sleep?

Author: Mbed OS decides it. I'll change it to "sleep or deep sleep" and add a [link](https://os.mbed.com/docs/mbed-os/v6.0-preview/apis/power-management-sleep.html)

> A major advantage of EventQueue over [Ticker/Timeout](#ticker-timeout) is that the handler is called in the same context (thread, in the case of RTOS) where the EventQueue is dispatched, thus ISR-related restrictions (e.g. no `printf`, no `Muxex` usage, etc.) do not apply.

Reviewer: ```suggestion
A major advantage of EventQueue over Ticker and Timeout is that the handler is called in the same context as the EventQueue is dispatched (thread, in the case of RTOS), so ISR-related restrictions (such as no `printf` oand no `Muxex`) do not apply.
```

Reviewer: Did you mean "mutex" or is "muxex" an alternative I'm not familiar with?

Author: It's a typo
